pkg/execution/pauses/block_leaser.go

```diff
@@ -0,0 +1,139 @@
+package pauses
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/redis/rueidis"
+)
+
+type redisBlockLeaser struct {
+	rc rueidis.Client
+
+	prefix   string
+	duration time.Duration
+}
+
+func (r redisBlockLeaser) indexKey(index Index) string {
+	return fmt.Sprintf("%s:%s:%s", r.prefix, index.WorkspaceID, index.EventName)
+}
+
+// Lease leases a given index, ensuring that only one worker can
+// flush an index at a time.
+func (r redisBlockLeaser) Lease(ctx context.Context, index Index) (leaseID ulid.ULID, err error) {
+	expires := time.Now().Add(r.duration)
+	leaseID, err = ulid.New(uint64(expires.UnixMilli()), rand.Reader)
+	if err != nil {
+		return leaseID, err
+	}
+
+	status, err := lease.Exec(
+		ctx,
+		r.rc,
+		[]string{r.indexKey(index)},
+		[]string{
+			strconv.Itoa(int(time.Now().UnixMilli())),
+			leaseID.String(),
+			"",
+			strconv.Itoa(int(r.duration.Seconds())),
+		},
+	).ToInt64()
+
+	switch status {
+	case -1:
+		return leaseID, err
+	default:
+		return leaseID, fmt.Errorf("unable to lease: already leased")
+	}
+}
+
+// Renew renews a lease while we are flushing an index.
+func (r redisBlockLeaser) Renew(ctx context.Context, index Index, existingLeaseID ulid.ULID) (newLeaseID ulid.ULID, err error) {
+	expires := time.Now().Add(r.duration)
+	newLeaseID, err = ulid.New(uint64(expires.UnixMilli()), rand.Reader)
+	if err != nil {
+		return newLeaseID, err
+	}
+
+	status, err := lease.Exec(
+		ctx,
+		r.rc,
+		[]string{r.indexKey(index)},
+		[]string{
+			strconv.Itoa(int(time.Now().UnixMilli())),
+			newLeaseID.String(),
+			existingLeaseID.String(),
+			strconv.Itoa(int(r.duration.Seconds())),
+		},
+	).AsInt64()
+	if err != nil {
+		return newLeaseID, fmt.Errorf("error renewing block lease: %w", err)
+	}
+	switch status {
+	case -1:
+		return newLeaseID, nil
+	default:
+		return newLeaseID, fmt.Errorf("unable to renew lease")
+	}
+}
+
+// Revoke drops a lease, allowing any other worker to flush an index.
+func (r redisBlockLeaser) Revoke(ctx context.Context, index Index, leaseID ulid.ULID) (err error) {
+	return r.rc.Do(ctx, r.rc.B().Del().Key(r.indexKey(index)).Build()).Error()
+}
+
+var (
+	lease = rueidis.NewLuaScript(`
+--[[
```

Comment:
let's just put this into an actual lua script file. still hard to read without syntax highlighting even if I'm used to it now.